torch/_decomp/__init__.py

```diff
@@ -250,13 +252,172 @@ def remove_decompositions(
 import torch._refs
 
 
+# Our strategy for deciding if we can preserve a op is following:
+# 1. The op should be known statically that it is functional
+# 2. If it is maybe aliasing, we decompose because we must know if an op
+#    is mutating or aliasing.
+# TODO (tmanlaibaatar) make this utility function and share it with functional_tensor
+# decomp part. (https://github.com/pytorch/pytorch/issues/129431)
+def _check_valid_to_preserve(op_overload):
+    if op_overload in FunctionalTensor.maybe_aliasing_or_mutating_ops:
+        return False
+    if op_overload in FunctionalTensor.metadata_fns:
+        return False
+
+    alias_info = len(
+        [i for i in op_overload._schema.arguments if i.alias_info is not None]
+    )
+
+    is_mutating_or_aliasing = alias_info != 0 or op_overload._schema.is_mutable
+
+    if is_mutating_or_aliasing:
+        return False
+
+    if not torch._C._dispatch_has_kernel(op_overload.name()):
+        return False
+
+    return True
+
+
+def _is_cia_op(op: "OpOverload") -> bool:
+    return (
+        torch._C._dispatch_has_kernel_for_dispatch_key(
+            op.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+        )
+        or torch._C.DispatchKey.CompositeImplicitAutograd in op.py_kernels
+    )
+
+
+def _is_special_op_to_decompose(*args, **kwargs):
+    kernel = kwargs["kernel"]
+    del kwargs["kernel"]
+    # Can't call kernel.decompose due to infinite recursion as
+    # we register this kernel to py_impl directly
+    dk = torch._C.DispatchKey.CompositeImplicitAutograd
+    if torch._C._dispatch_has_kernel_for_dispatch_key(
+        kernel.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+    ):
+        return kernel._op_dk(dk, *args, **kwargs)
+    else:
+        return NotImplemented
```

Comment:
nit: shouldn't there be an invariant here that you **never get here**? You are carefully only adding this decomp into the decomp table for ops that you have promised have a CIA impl.

(and if you need to land this sooner due to timeline pressure that's fine, but this seems worth rationalizing)

Reply:
Ahh i just copy pasted it from func.decompose. But yeah i think you are right, this should be an assert.